<a href="https://colab.research.google.com/github/Astrid-Itzel/data_analysis_ECOBICI/blob/main/Proyecto_ECOBICI_Ciencia_de_datos_Tecnolochicas_AIAZ.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **Proyecto final - ECOBICI**

## **Introducción**

Los sistemas de transporte representan para las grandes ciudades una herramienta importante para mejorar la movilidad urbana, reducir la congestión vehicular y promover alternativas de movilidad más sustentables. Resulta necesario impulsar soluciones de transporte que permitan a la población desplazarse de manera eficiente, accesible y amigable con el medio ambiente, especialmente en ciudades con alta densidad poblacional y grandes retos de movilidad.

En la CDMX, el sistema ECOBICI se ha consolidado como una alternativa de transporte que permite a miles de personas realizar trayectos diarios de manera rápida y eficiente, reduciendo los tiempos que el tráfico automovilistico representa y las incomodidades de metros y camiones saturados.

La operación de un sistema de bicicletas compartidas implica diversos retos relacionados con la distribución de unidades, la disponibilidad de bicicletas en las estaciones, los patrones de movilidad de los usuarios y la capacidad de adaptación del servicio a las necesidades de la población.

El análisis de los datos generados por ECOBICI permite identificar patrones de comportamiento, tendencias de uso y posibles áreas de oportunidad para mejorar la eficiencia del sistema. Mediante técnicas de exploración, limpieza y procesamiento de datos es posible estudiar variables como horarios de mayor demanda, estaciones con mayor flujo de bicicletas, duración promedio de los viajes, características demográficas de los usuarios y diferencias en los patrones de movilidad durante distintos días de la semana.

La importancia de este análisis radica en que los datos pueden convertirse en información útil para la toma de decisiones. Por ejemplo, identificar estaciones con alta demanda podría ayudar a optimizar la redistribución de bicicletas y evitar que ciertos puntos se queden sin unidades disponibles, mientras que detectar estaciones con poca actividad permitiría evaluar si requieren estrategias de promoción, ajustes operativos o una posible reorganización del servicio. Asimismo, conocer el perfil de los usuarios puede contribuir al diseño de campañas de comunicación más específicas, orientadas a incentivar el uso de ECOBICI en sectores de la población con menor participación.

## **Preguntas de investigación**

A través del análisis exploratorio de datos se busca responder preguntas relacionadas con el funcionamiento del sistema, tales como:

¿Cuáles son las estaciones con mayor número de retiros y arribos de bicicletas?

*   ¿Qué estaciones presentan mayor demanda y podrían requerir una redistribución más frecuente de unidades?
*   ¿Existen estaciones donde las bicicletas permanecen con poca movilidad durante el periodo analizado?
*   ¿En qué horarios se concentra la mayor cantidad de viajes y cómo varía la demanda durante la semana?
*   ¿Cuánto tiempo duran, en promedio, los viajes realizados por los usuarios?
*   ¿Existen diferencias en los patrones de uso entre días laborales y fines de semana?
*   ¿Cuál es el perfil general de los usuarios del sistema considerando variables como edad y género?
*   ¿Qué grupos de usuarios tienen mayor participación dentro del servicio?
*   ¿Qué oportunidades existen para aumentar la adopción del sistema en otros segmentos de la población?

Los resultados obtenidos pueden servir como base para proponer mejoras operativas, como una planeación más eficiente de la distribución de bicicletas, la identificación de horarios críticos, la optimización de estaciones estratégicas y el desarrollo de estrategias de comunicación enfocadas en las necesidades reales de los usuarios. De esta manera, el análisis de datos permite transformar los registros de viajes en conocimiento útil para fortalecer un sistema de movilidad urbana más eficiente, accesible y sostenible.
## **Hipótesis**

Se espera encontrar que la demanda de ECOBICI presenta patrones asociados principalmente con horarios laborales, ubicación de estaciones cercanas a zonas de alta actividad y características específicas de los usuarios, lo que permitiría identificar oportunidades para mejorar la distribución de bicicletas y las estrategias de promoción del servicio.

## **Análisis**

El análisis se realizará sobre los siguiente puntos:

*  Redistribución operativa: estaciones que se vacían por la mañana y se llenan por la tarde (por ejemplo, zonas residenciales vs zonas laborales).
*  Horas pico: detectar ventanas donde debería haber más personal redistribuyendo bicicletas.
*  Mapeo de estaciones críticas: estaciones origen/destino más utilizadas.
*  Perfil del usuario: edad promedio, distribución por género, usuarios predominantes.
*  Patrones temporales: diferencias entre lunes-viernes y fines de semana.
*  Movilidad urbana: rutas más frecuentes entre estaciones.
*  Predicción de demanda (como extensión futura): estimar cuántas bicicletas necesitará una estación en cierto horario.

## **Importación de librerías**

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
pd.set_option('display.float_format', lambda x: '%.2f' % x) # mostrará el display de floats con únicamente 2 decimales

## **1. IMPORTACIÓN Y MANEJO DE DATOS**

## **Subir la base de datos a Google Colab**

Para la presente investigación se utilizará la base de datos real de ECOBICI, durante el periodo de enero de 2026. La cual puede ser descargada de: https://ecobici.cdmx.gob.mx/datos-abiertos/

Asimismo, se creó un segundo data frame con la información relacionada a la ubicación de las estaciones de ECOBICI, obtenida en: https://ecobici.cdmx.gob.mx/mapa/
Dicha información fue descargada y guardada a modo de archivo csv para porteriormente su carga y manipulación como data frame.

Luego de haber sido descargada, se ejecutará la función a continuación para subir el archivo CSV.

In [ ]:
from google.colab import files

files.upload()

Saving 2026-01.csv to 2026-01.csv
Buffered data was truncated after reaching the output size limit.

## **Guardar como dataframe la BD en Pandas**

Leemos el archivo CSV con la BD y lo guardamos en la variable df (por convención para data frame).

El método **read_csv()** de Pandas lee el archivo y devuelve un objeto de tipo DataFrame.

In [ ]:
df = pd.read_csv("2026-01.csv") #BD principal de ECOBICI enero 2026

df_estaciones = pd.read_csv("estaciones_ecobici.csv") #BD secundaria para conocer ubicaciones de las estaciones de ECOBICI


## **Revisión de la BD**

Durante el análisis inicial del conjunto de datos de ECOBICI correspondiente a enero de 2026, se identificó que la BD está conformada por 1,525,302 de registros de viajes y 9 columnas:

1.   Genero_Usuario
2.   Edad_Usuario
3.   Bici
4.   Ciclo_Estacion_Retiro
5.   Fecha_Retiro
6.   Hora_Retiro
7.   Ciclo_EstacionArribo
8.   Fecha_Arribo
9.   Hora_Arribo

Nota:
*   Las fechas y horas están almacenadas como texto (object) y deberán transformarse para realizar análisis.


In [ ]:
print("Número de filas y columnas: \n")
print(df.shape) #Para obtener número de filas y columnas (filas, columnas)

print("-"*100)
print("Construcción de la tabla: \n")
df.info() # Responde cómo está construida la tabla, incluyendo los tipos de datos, nombres de las columnas y lacantidad de registros Non-Null

print("-"*100)
print("Ejemplo de las primeras filas: \n")
display(df.head()) # Muestra las primeras filas de un DataFrame


Número de filas y columnas: 

(1525302, 9)
----------------------------------------------------------------------------------------------------
Construcción de la tabla: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1525302 entries, 0 to 1525301
Data columns (total 9 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   Genero_Usuario         1525301 non-null  object 
 1   Edad_Usuario           1525223 non-null  float64
 2   Bici                   1525302 non-null  int64  
 3   Ciclo_Estacion_Retiro  1525302 non-null  object 
 4   Fecha_Retiro           1525302 non-null  object 
 5   Hora_Retiro            1525302 non-null  object 
 6   Ciclo_EstacionArribo   1525302 non-null  object 
 7   Fecha_Arribo           1525302 non-null  object 
 8   Hora_Arribo            1525302 non-null  object 
dtypes: float64(1), int64(1), object(7)
memory usage: 104.7+ MB
----------------------------------------------------------

,Genero_Usuario,Edad_Usuario,Bici,Ciclo_Estacion_Retiro,Fecha_Retiro,Hora_Retiro,Ciclo_EstacionArribo,Fecha_Arribo,Hora_Arribo
0,M,30.00,4933404,059,31/12/2025,23:25:20,565,01/01/2026,00:00:16
1,M,27.00,4162376,180,31/12/2025,23:43:53,672,01/01/2026,00:00:31
2,M,29.00,3749272,014,31/12/2025,23:52:15,030,01/01/2026,00:00:50
3,M,44.00,5680772,014,31/12/2025,23:46:43,145,01/01/2026,00:00:53
4,M,22.00,5128459,031,31/12/2025,23:39:58,155,01/01/2026,00:01:05


## **`Datos estadísticos`**

A partir del análisis descriptivo se observa que la edad promedio de los usuarios es de aproximadamente 34.57 años, con una edad mínima registrada de 16 años y una máxima de 100 años. La distribución de edades muestra que el 50% de los usuarios tiene 33 años o menos, mientras que el 75% se encuentra por debajo de los 40 años, esto indica una mayor concentración de usuarios jóvenes y adultos. Asimismo, se identificó que el género más frecuente corresponde a "M", con más de un millón de registros, aunque existe una pequeña cantidad de valores faltantes en las variables de género y edad. Estos resultados permiten obtener un panorama general de los usuarios y detectar aspectos que deberán considerarse durante la etapa de limpieza de los datos.

**Datos estadísticos numéricos**


In [10]:
descripcion_tabla = df.describe() # describe() por defecto analiza únicamente columnas numéricas, por lo que únicamente muestra  Edad_usuario y Bici
print(descripcion_tabla)

#utilizamos print para no generar una tabla, pues al ejecutar df.describe() Pandas creaba un DataFrame temporal con las estadísticas y Colab intentaba mostrar ese DataFrame como una tabla visual, lo que generaba error por la cantidad de registros en la BD.

       Edad_Usuario       Bici
count    1525223.00 1525302.00
mean          34.57 5472382.29
std            9.82 2029855.42
min           16.00 2000461.00
25%           28.00 3739224.00
50%           33.00 5482971.00
75%           40.00 7229172.00
max          100.00 8999708.00


**Datos estadísticos de objetos**


In [11]:
display(df.describe(include="object")) # (include="all") o (include="object") mostrarían todas las columnas o las columnas que no son numéricas, respectivamente.

#count → cantidad de valores no nulos
#unique → cuántos valores diferentes existen
#top → el valor que más se repite
#freq → cuántas veces aparece

,Genero_Usuario,Ciclo_Estacion_Retiro,Fecha_Retiro,Hora_Retiro,Ciclo_EstacionArribo,Fecha_Arribo,Hora_Arribo
count,1525301,1525302,1525302,1525302,1525302,1525302,1525302
unique,4,677,42,69706,677,31,70772
top,M,548,28/01/2026,18:06:18,271-272,28/01/2026,18:28:59
freq,1029960,10359,64546,66,13536,64547,61


## **Calidad de los datos -  isnull()**

In [12]:
# Contamos los valores faltantes por columna y mandamos a imprimir con print, ya que de esta manera df.isnull().sum() se intenta generar una tabla html que por la cantidad de registros Google colab tiene dificultad en generar

nulos = df.isnull().sum()
display(nulos)

,0
Genero_Usuario,1
Edad_Usuario,79
Bici,0
Ciclo_Estacion_Retiro,0
Fecha_Retiro,0
Hora_Retiro,0
Ciclo_EstacionArribo,0
Fecha_Arribo,0
Hora_Arribo,0


 **2. LIMPIEZA DE DATOS**

**Valores faltantes**



De acuerdo con la información recabada hasta el momento, es importante tomar en consideración los siguientes puntos con respecto a los valores faltantes:

*   existe 1 valor nulo en Genero_Usuario.
*   existen 79 valores nulos en Edad_Usuario.

La presencia de valores faltantes únicamente contempla las variables Genero_Usuario y Edad_Usuario. Debido a que representan una cantidad mínima respecto al total de registros de toda la BD, se evaluó la estrategia adecuada para su tratamiento sin afectar significativamente el volumen de información disponible.

Durante la revisión de los valores faltantes en la variable Edad_Usuario, se identificaron 79 registros sin información de edad, lo que representa una proporción mínima respecto al total de observaciones del conjunto de datos. Debido a que estos registros contienen información relevante sobre los viajes realizados, como estaciones de origen y destino, horarios y bicicletas utilizadas, se decidió conservarlos en lugar de eliminarlos, evitando perder información útil para análisis posteriores relacionados con los patrones de movilidad dentro del sistema ECOBICI.

Para completar los valores faltantes de edad se aplicó una estrategia de imputación utilizando la mediana de la variable, en lugar de la media. Esta decisión se tomó debido a que la edad presenta un rango amplio de valores, incluyendo posibles edades extremas, y la mediana es una medida estadística menos sensible a estos valores atípicos. De esta manera, se mantiene una representación más adecuada del usuario promedio sin alterar significativamente la distribución original de los datos.




*  **Edades faltantes**

**Rellenar edades faltantes con la mediana**

In [13]:
df_clean = df.copy() #Hacemos una copia por seguridad para guardar la BD limpia

mediana_edad = df_clean["Edad_Usuario"].median() # encontramos la mediana en la columna "Edad_Usuario" del df_clean y lo guardamos en la variable mediana_edad. Pandas revisa todos los valores de la columna Edad_Usuario, los ordena y obtiene el valor central.
print(mediana_edad)

df_clean["Edad_Usuario"] = df["Edad_Usuario"].fillna(mediana_edad) #En la columna Edad_Usuario, donde se encuentre valores vacíos, coloca la mediana. fillna() no modifica la columna original automáticamente, por eso es necesario ejecturar df["Edad_Usuario"] = ... para guardar el resultado nuevamente en la columna.

33.0


**Verificar que ya no existan valores nulos en Edad_Usuario**

In [14]:
print(f"Edades nulas en df_clean: {df_clean['Edad_Usuario'].isnull().sum()}") #En la columna "Edad_Usuario" del df_clean ejecuta isnull().sum() para sumar los valores nulos

print("Edades nulas en df:", df['Edad_Usuario'].isnull().sum()) #A diferencia del df_clean, el original df continúa con valores nulos


Edades nulas en df_clean: 0
Edades nulas en df: 79


*  **Género faltante**

De 1,525,302 registros, solamente 1 registro no tiene información de género. Debido a que representa un porcentaje insignificante respecto al total de registros y que no existe una forma objetiva de determinar la categoría correcta sin introducir información artificial, se decidió eliminar éste. Esta acción permitió mantener únicamente datos completos para los análisis relacionados con características demográficas de los usuarios.

In [15]:
df_clean = df_clean.dropna(subset=["Genero_Usuario"]) #Esto elimina únicamente las filas donde Genero_Usuario está vacío, sin afectar otras columnas.

#dropna(): elimina las filas o columnas que contienen valores nulos (NaN) en un DataFrame.
#subset: indica las columnas específicas que dropna() debe revisar para encontrar valores nulos.

Verificamos si se eliminó el registro nulo de Genero_Usuario

In [16]:
print(df_clean["Genero_Usuario"].isnull().sum())

#El resultado debe ser 0 ya que no debe existir más valores nulos

0


* **Edades atípicas**

Durante el análisis exploratorio de la variable Edad_Usuario, se identificaron registros con edades elevadas, incluyendo valores superiores a los 90 años. Aunque estas edades son biológicamente posibles, representan casos poco frecuentes dentro del contexto de un sistema de bicicletas compartidas, por lo que se consideraron valores potencialmente atípicos que requieren una revisión adicional.

Debido a que la base de datos corresponde a registros de viajes y no a usuarios únicos, la frecuencia observada representa la cantidad de viajes realizados por personas con determinada edad registrada, y no necesariamente el número de usuarios. Por esta razón, se decidió conservar estos registros para evitar la pérdida de información, pero se creó la variable auxiliar Edad_Atipica, la cual permite identificar y analizar por separado aquellos viajes asociados a edades iguales o superiores a 90 años. Esta clasificación facilita evaluar el impacto de estos valores en los análisis posteriores sin asumir automáticamente que se trata de errores de captura.

In [17]:
df_clean[df_clean["Edad_Usuario"] >= 90].shape[0] #Cuántos viajes fueron realizados por personas mayores a 90 años

df_clean[df_clean["Edad_Usuario"] >= 90]["Edad_Usuario"].value_counts().sort_index() # Suma de viajes por edad, donde la edad es mayor a 90

,count
Edad_Usuario,
90.00,3
96.00,12
100.00,6


In [18]:
df_clean["Edad_Atipica"] = df_clean["Edad_Usuario"] >= 90 #crea una nueva columna llamada "Edad_Atipica" que tendrá valores booleanos revisando si el valor de "Edad._Usuario" es igual o mayor a 90 años

In [19]:
print(df_clean["Edad_Atipica"].value_counts()) # devuelve la cuenta de True y False

Edad_Atipica
False    1525280
True          21
Name: count, dtype: int64


In [20]:
display(df_clean[df_clean["Edad_Atipica"]]["Edad_Usuario"].value_counts().sort_index()) #devuelve la suma de True por edad

,count
Edad_Usuario,
90.00,3
96.00,12
100.00,6


**Revisión de registros duplicados**



In [21]:
df_clean.duplicated().sum() #suma el total de reistros duplicados en df_clean

np.int64(0)

**Revisión de formato de datos**

* **Actualización a formato de datetime**

Durante la etapa de limpieza y preparación de datos, se realizó la conversión de las columnas Fecha_Retiro y Fecha_Arribo del tipo de dato object a formato de fecha (datetime). Aunque los valores contenían información de fecha, originalmente estaban almacenados como String, lo que limitaba la capacidad de realizar operaciones temporales y análisis más complejos sobre los registros de viajes.

Este cambio fue necesario para garantizar que las fechas fueran interpretadas correctamente por las herramientas de análisis, permitiendo realizar cálculos y agrupaciones relacionadas con el comportamiento temporal de los viajes, como identificar patrones por día, analizar periodos de mayor demanda o validar la consistencia entre las fechas de retiro y arribo. Para evitar errores de interpretación, se especificó el formato original de los datos (día/mes/año), asegurando una conversión adecuada.

In [22]:
df_clean.dtypes #.dtypes revisa el tipo de datos de todas las columnas

,0
Genero_Usuario,object
Edad_Usuario,float64
Bici,int64
Ciclo_Estacion_Retiro,object
Fecha_Retiro,object
Hora_Retiro,object
Ciclo_EstacionArribo,object
Fecha_Arribo,object
Hora_Arribo,object
Edad_Atipica,bool


In [23]:
df_clean["Fecha_Retiro"] = pd.to_datetime(
    df_clean["Fecha_Retiro"],
    format="%d/%m/%Y"
)
 # pd.to_datetime esta es una función de Pandas que convierte valores a formato fecha.
 # format="%d/%m/%Y" indica a Pandas cómo está escrita la fecha, esto es necesario porque por default Pandas lee mes/día/año

df_clean["Fecha_Arribo"] = pd.to_datetime(
    df_clean["Fecha_Arribo"],
    format="%d/%m/%Y"
)


**Cambio de formato de edades**

Durante el análisis de la información, se comprobó que las edades de los usuarios existen en formato float aunque no tienen realmente decimales, por lo que se tomó la decisión de cambiar el formato a int. Al tratarse de una variable que representa años cumplidos, el formato entero resulta más adecuado para su interpretación y análisis.

In [24]:
df_clean["Edad_Usuario"] = df_clean["Edad_Usuario"].astype(int)

df_clean["Edad_Usuario"].dtype

dtype('int64')

**Valores de género de los usuarios**

Durante la revisión de la variable Genero_Usuario se identificaron cuatro categorías: M, F, O y ?. Las categorías M y F corresponden a los valores predominantes de la variable, mientras que O corresponde a Otros, y el valor ? fue identificado como un dato no especificado o desconocido, por lo que fue recodificado como "No especificado" para mejorar la interpretación de la información sin eliminar registros de viajes.

In [25]:
df_clean["Genero_Usuario"].value_counts() #realiza la cuenta de cada tipo de género (masculino, femenino, otro y desconocido)

,count
Genero_Usuario,
M,1029960
F,418009
?,49853
O,27479


In [26]:
df_clean["Genero_Usuario"] = df_clean["Genero_Usuario"].replace("?", "No especificado") #remplazamos los "?" por "No especificado" para mayor claridad

### Resumen de cambios realizados durante la limpieza de datos

* Se creó una copia de trabajo (df_clean) a partir de la base original, con el objetivo de realizar modificaciones sin alterar los datos fuente.

* Se realizó una revisión general de valores nulos en las variables de la base de datos.

* En la variable Edad_Usuario:

  * Se identificaron registros con valores faltantes.
  * Los valores nulos fueron reemplazados utilizando la mediana de la variable, debido a que esta medida es menos sensible a valores extremos y permite conservar la distribución general de las edades.
  * Se modificó el tipo de dato de float64 a int, debido a que los valores registrados no contenían decimales significativos y la variable representa años cumplidos.

* En la variable Genero_Usuario:

  * Se identificó un registro sin información de género y se eliminó al no contar con información suficiente para determinar un valor válido.
  * Se detectó la categoría ?, la cual fue interpretada como un valor no especificado y se recodificó como "No especificado" para mejorar la interpretación de los datos sin perder registros.
  * Se conservaron las categorías existentes (M, F y O) debido a que representan valores registrados en la fuente original.

* Se realizó una revisión de registros duplicados:

  * No se encontraron filas duplicadas dentro de la base de datos, por lo que no fue necesario eliminar registros repetidos.

* Se realizó una revisión de valores atípicos en la variable Edad_Usuario:

  * Se identificaron edades iguales o superiores a 90 años como valores potencialmente atípicos.
  * Se creó la variable auxiliar Edad_Atipica para identificar estos casos sin eliminarlos, ya que, aunque son poco frecuentes, son edades plausibles dentro del contexto de un sistema de bicicletas compartidas.
  * Se consideró que la base contiene registros por viaje y no usuarios únicos, por lo que la frecuencia observada representa viajes realizados y no cantidad de personas.


## 3.  Análisis exploratorio

Una vez concluida la etapa de limpieza de datos, se presenta la versión actualizada de la base de datos (df_clean), la cual será utilizada para realizar el análisis exploratorio. Esta nueva versión contiene los ajustes necesarios para mejorar la calidad y consistencia de la información, incluyendo el tratamiento de valores nulos, revisión de registros duplicados, transformación de tipos de datos y creación de variables auxiliares que facilitarán el análisis posterior.

El análisis exploratorio de datos tiene como objetivo conocer las características principales de la base resultante, identificar patrones de comportamiento y obtener una primera interpretación de la información disponible. Para ello, se aplicarán técnicas de exploración y estadística descriptiva utilizando herramientas de Pandas, con el fin de analizar la distribución de variables, la frecuencia de los registros, el comportamiento temporal de los viajes y las características generales de los usuarios.

La base de datos utilizada corresponde a los viajes registrados por el sistema Ecobici durante enero de 2026. Es importante considerar que cada registro representa un viaje realizado y no un usuario individual, por lo que los resultados obtenidos reflejan patrones de movilidad y uso del sistema durante el periodo analizado.

In [27]:
df_clean.shape # El número de filas y columnas en el data frame limpio corresponde a 1525302 filas y 13 columnas

(1525301, 10)

In [28]:
df_clean.info() #muestra las columnas, que ya no secuenta con datos null en ninguna columna o fila y los tipos de datos actualizados

<class 'pandas.core.frame.DataFrame'>
Index: 1525301 entries, 0 to 1525301
Data columns (total 10 columns):
 #   Column                 Non-Null Count    Dtype         
---  ------                 --------------    -----         
 0   Genero_Usuario         1525301 non-null  object        
 1   Edad_Usuario           1525301 non-null  int64         
 2   Bici                   1525301 non-null  int64         
 3   Ciclo_Estacion_Retiro  1525301 non-null  object        
 4   Fecha_Retiro           1525301 non-null  datetime64[ns]
 5   Hora_Retiro            1525301 non-null  object        
 6   Ciclo_EstacionArribo   1525301 non-null  object        
 7   Fecha_Arribo           1525301 non-null  datetime64[ns]
 8   Hora_Arribo            1525301 non-null  object        
 9   Edad_Atipica           1525301 non-null  bool          
dtypes: bool(1), datetime64[ns](2), int64(2), object(5)
memory usage: 117.8+ MB


In [29]:
display(df_clean.describe()) #muestra las estadísticas del df limpio

,Edad_Usuario,Bici,Fecha_Retiro,Fecha_Arribo
count,1525301.00,1525301.00,1525301,1525301
mean,34.57,5472381.99,2026-01-17 02:42:33.229820928,2026-01-17 02:46:30.797095168
min,16.00,2000461.00,2025-08-02 00:00:00,2026-01-01 00:00:00
25%,28.00,3739224.00,2026-01-10 00:00:00,2026-01-10 00:00:00
50%,33.00,5482971.00,2026-01-17 00:00:00,2026-01-17 00:00:00
75%,40.00,7229172.00,2026-01-24 00:00:00,2026-01-24 00:00:00
max,100.00,8999708.00,2026-01-31 00:00:00,2026-01-31 00:00:00
std,9.82,2029856.05,NaN,NaN


**Perfil de usuario**

**Género**

El análisis demográfico muestra una marcada predominancia del género masculino, el cual concentra el 67.53% del total de los viajes realizados en el mes. Por su parte, las mujeres representan el 27.41% de la actividad del sistema. El 5.07% restante se distribuye entre los viajes correspondientes a la categoría "No especificado" (3.27%), que agrupa los registros antes codificados con un signo de interrogación, y la opción "Otro" (1.80%). Esta brecha resalta un área de oportunidad para el diseño de políticas públicas que promuevan una mayor inclusión y adopción del servicio entre las mujeres.

In [79]:
# Cambiar las letras por los nombres completos para el análisis exploratorio
df_clean["Genero_Usuario"] = df_clean["Genero_Usuario"].replace({
    "M": "Masculino",
    "F": "Femenino",
    "O": "Otros"
})

In [80]:
display("Porcentaje por Género:")
df_clean['Genero_Usuario'].value_counts(normalize=True) * 100 #muestra qué porcentaje existe enrelación a cada género

'Porcentaje por Género:'

,proportion
Genero_Usuario,
Masculino,67.53
Femenino,27.41
No especificado,3.27
Otros,1.80


**Edad**

La edad promedio del usuario de ECOBICI es de 34.57 años. El análisis muestra que el 50% de las personas tiene 33 años o menos (mediana) y el 75% es menor de 40 años, lo que confirma que el servicio es utilizado predominantemente por adultos jóvenes en edad universitaria y laboral. Aunque el rango abarca desde los 16 hasta un valor extremo de 100 años, la concentración de los datos refleja que el sistema se consolida como una solución de movilidad para los sectores demográficos más activos.

In [31]:
print("\nResumen estadístico de la Edad:")
df_clean['Edad_Usuario'].describe()


Resumen estadístico de la Edad:


,Edad_Usuario
count,1525301.00
mean,34.57
std,9.82
min,16.00
25%,28.00
50%,33.00
75%,40.00
max,100.00


**Exploración de horarios**

El análisis de los viajes revela que el sistema ECOBICI en la Ciudad de México experimenta su mayor volumen de actividad durante la tarde. La hora de máxima demanda absoluta ocurre a las 18:00 horas (6:00 PM), registrando más de 133,000 viajes en el mes, seguida de cerca por las 17:00 horas con 123415 viajes. Esta marcada concentración coincide directamente con el horario habitual de salida de las zonas laborales corporativas y escolares, consolidando a ECOBICI como una herramienta clave para evitar el congestionamiento vehicular.

Por otro lado, la mañana presenta su propio pico de actividad concentrado a las 08:00 AM. Sin embargo, resulta interesante observar que el volumen de viajes en este horario es menor en comparación con el registrado por la tarde, registrando 107995 viajes. Este comportamiento sugiere que los traslados matutinos de los usuarios hacia sus destinos suelen estar más dispersos cronológicamente o se concentran en zonas muy específicas, mientras que el regreso y la movilidad vespertina saturan el sistema de manera más uniforme y masiva. Estos datos aportan una base sólida para que el equipo operativo planifique de manera estratégica la redistribución de unidades, priorizando el reabastecimiento de estaciones justo antes de las 5:00 PM.

In [85]:
# Convertimos la variable Hora_Retiro al formato reconocido por Pandas
df_clean["Hora_Retiro"] = pd.to_datetime(
    df_clean["Hora_Retiro"],
    format="%H:%M:%S"
)

# Extraemos la hora del viaje como variable auxiliar para análisis posteriores
df_clean["Hora_Entera"] = df_clean["Hora_Retiro"].dt.hour

# 4. Manipulación de datos


Durante la etapa de preparación de datos, se realizó la conversión de las columnas Hora_Retiro y Hora_Arribo de texto (object) a un formato temporal reconocido por Pandas. Este cambio fue necesario porque, aunque los valores contenían información de hora, al estar almacenados como cadenas de texto no podían utilizarse directamente para realizar operaciones de tiempo, como calcular diferencias entre horarios, identificar periodos de mayor demanda o analizar patrones de uso por hora. La conversión permite que estas variables sean interpretadas correctamente por las herramientas de análisis y facilita la generación de nuevas métricas relacionadas con el comportamiento temporal de los viajes.

Adicionalmente, se propone la creación de las columnas derivadas Inicio_Viaje, Fin_Viaje, Duracion_Minutos y Dia_Semana, sin eliminar las variables originales de fecha y hora. Aunque la base ya cuenta con la fecha y hora de retiro y arribo por separado, la combinación de ambas dimensiones permite representar de manera completa el momento exacto en que inicia y termina cada viaje. Esto facilita el cálculo de la duración de los recorridos y permite realizar análisis más avanzados sobre tiempos de traslado, eficiencia del servicio y patrones de movilidad. Las columnas originales se conservan para mantener la información de origen y las nuevas variables funcionan como indicadores calculados para el análisis.

En este apartado se comenzarán a responder algunas de las preguntas de investigación planteadas.

## ** Estaciones con mayor y menor demanda**

Debido a que en la base de datos descargada de ECOBICI refiere a las estaciones con su número id, se utilizó una segunda BD para obtener la dirección exacta de éstas.  

In [34]:
#REVISIÓN DE LA SEGUNDA BD. df_estaciones
df_estaciones.head()


,numero_estacion,direccion
0,001,RÍO SENA-RÍO BALSAS
1,002,RÍO GUADALQUIVIR - RÍO NAZAS
2,003,REFORMA - INSURGENTES
3,004,RÍO NILO - RÍO PANUCO
4,005,RÍO PÁNUCO RÍO TIBER


In [35]:
df_estaciones.info() #el número de estación se maneja en string debido a que no sólo hay números sino rangos por las estaciones que comparten ubicación

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 677 entries, 0 to 676
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   numero_estacion  677 non-null    object
 1   direccion        677 non-null    object
dtypes: object(2)
memory usage: 10.7+ KB


In [36]:
# Creamos una columna extra que indique la dirección de origen, a partir del id de la estación, uniendo los dos data frames
df_clean = pd.merge(
    df_clean,
    df_estaciones,
    left_on='Ciclo_Estacion_Retiro',
    right_on='numero_estacion',
    how='left'
)

# Renombrar la columna nueva dentro de df_clean para no confundir la dirección del arribo
df_clean = df_clean.rename(columns={'direccion': 'direccion_origen'})


In [37]:
# Unir para obtener los datos de la estación de destino (Arribo) directamente en df_clean
df_clean = pd.merge(
    df_clean,
    df_estaciones,
    left_on='Ciclo_EstacionArribo',   # columna destino en df_clean
    right_on='numero_estacion',       # columna de ID en el df_estaciones
    how='left'
)

# Renombrar la nueva columna "direccion" para saber que es el arribo
df_clean = df_clean.rename(columns={'direccion': 'direccion_arribo'})


In [38]:
# Eliminar las columnas repetidas _x y _y
df_clean = df_clean.drop(columns=['numero_estacion_x', 'numero_estacion_y'])

In [39]:
print(df_clean.head())

  Genero_Usuario  Edad_Usuario     Bici Ciclo_Estacion_Retiro Fecha_Retiro  \
0              M            30  4933404                   059   2025-12-31   
1              M            27  4162376                   180   2025-12-31   
2              M            29  3749272                   014   2025-12-31   
3              M            44  5680772                   014   2025-12-31   
4              M            22  5128459                   031   2025-12-31   

  Hora_Retiro Ciclo_EstacionArribo Fecha_Arribo Hora_Arribo  Edad_Atipica  \
0    23:25:20                  565   2026-01-01    00:00:16         False   
1    23:43:53                  672   2026-01-01    00:00:31         False   
2    23:52:15                  030   2026-01-01    00:00:50         False   
3    23:46:43                  145   2026-01-01    00:00:53         False   
4    23:39:58                  155   2026-01-01    00:01:05         False   

   Hora_Entera           direccion_origen                  direccion

* **Las 10 estaciones de origen más utilizadas**

In [40]:
print("\nLas 10 estaciones con mayor número de retiros:")

# Agrupamos por ID y dirección, y contamos cuántos viajes tiene cada combinación
top_mayor_devoluciones = df_clean.groupby(['Ciclo_Estacion_Retiro', 'direccion_origen']).size()

# 2. Ordenamos los valores de menor a mayor y tomamos los primeros 10
print(top_mayor_devoluciones.sort_values(ascending=False).head(10))


Las 10 estaciones con mayor número de retiros:
Ciclo_Estacion_Retiro  direccion_origen                        
548                    DR. MARIANO AZUELA - JOSE ANTONIO ALZATE    10359
208                    HESIODO-LAMARTINE                            9934
027                    REFORMA- HAVRE                               8348
064                    SONORA - ÁMSTERDAM                           7859
271-272                JESÚS GARCÍA - CARLOS J. MENESES             7711
555                    RIBERA DE SAN COSME - INSURGENTES CENTRO     7325
038                    COZUMEL - PUEBLA                             7103
014                    REFORMA - RÍO DE PLATA                       6736
031                    HAMBURGO - INSURGENTES                       6355
018                    REFORMA - RÍO RHIN                           6335
dtype: int64


* **Las 10 estaciones de arribo más utilizadas**

In [41]:
print("\nLas 10 estaciones con mayor número de devoluciones:")

# Agrupamos por ID y dirección, y contamos cuántos viajes tiene cada combinación
top_mayor_devoluciones = df_clean.groupby(['Ciclo_EstacionArribo', 'direccion_arribo']).size()

# 2. Ordenamos los valores de menor a mayor y tomamos los primeros 10
print(top_mayor_devoluciones.sort_values(ascending=False).head(10))


Las 10 estaciones con mayor número de devoluciones:
Ciclo_EstacionArribo  direccion_arribo                        
271-272               JESÚS GARCÍA - CARLOS J. MENESES            13536
548                   DR. MARIANO AZUELA - JOSE ANTONIO ALZATE    12267
027                   REFORMA- HAVRE                               8806
064                   SONORA - ÁMSTERDAM                           8409
208                   HESIODO-LAMARTINE                            8300
014                   REFORMA - RÍO DE PLATA                       8039
555                   RIBERA DE SAN COSME - INSURGENTES CENTRO     7378
038                   COZUMEL - PUEBLA                             6782
134                   ÁLVARO OBREGÓN-ORIZABA                       6673
029                   REFORMA - BUCARELI                           6667
dtype: int64


* **Las 10 estaciones de origen menos utilizadas**

In [42]:
print("\nLas 10 estaciones con menor número de retiros:")

# Agrupamos por ID y dirección, y contamos cuántos viajes tiene cada combinación
top_menor_retiros = df_clean.groupby(['Ciclo_Estacion_Retiro', 'direccion_origen']).size()

# 2. Ordenamos los valores de menor a mayor y tomamos los primeros 10
print(top_menor_retiros.sort_values(ascending=True).head(10))


Las 10 estaciones con menor número de retiros:
Ciclo_Estacion_Retiro  direccion_origen                     
687                    FRANCISCO SOSA - PANZACOLA               288
664                    HORTENSIA - 2DA CERRADA DE MINERVA       332
666                    MORAS - NARANJO                          332
662                    MANZANO - CIRCUITO INTERIOR              334
665                    MORAS - MANGO                            340
707                    IRLANDA - DIVISIÓN DEL NORTE             426
694                    IGNACIO ALDAMA - CENTENARIO              470
650                    XICOTÉNCATL- PROLONGACIÓN XICOTÉNCATL    490
516                    TEXCOCO - NILO                           499
443                    BRUNO TRAVEN-GOLONDRINAS                 499
dtype: int64


* **Las 10 estaciones de arribo menos utilizadas**

In [43]:
print("\nLas 10 estaciones con menor número de devoluciones:")

# Agrupamos por ID y dirección, y contamos cuántos viajes tiene cada combinación
top_menor_devoluciones = df_clean.groupby(['Ciclo_EstacionArribo', 'direccion_arribo']).size()

# 2. Ordenamos los valores de menor a mayor y tomamos los primeros 10
print(top_menor_retiros.sort_values(ascending=True).head(10))


Las 10 estaciones con menor número de devoluciones:
Ciclo_Estacion_Retiro  direccion_origen                     
687                    FRANCISCO SOSA - PANZACOLA               288
664                    HORTENSIA - 2DA CERRADA DE MINERVA       332
666                    MORAS - NARANJO                          332
662                    MANZANO - CIRCUITO INTERIOR              334
665                    MORAS - MANGO                            340
707                    IRLANDA - DIVISIÓN DEL NORTE             426
694                    IGNACIO ALDAMA - CENTENARIO              470
650                    XICOTÉNCATL- PROLONGACIÓN XICOTÉNCATL    490
516                    TEXCOCO - NILO                           499
443                    BRUNO TRAVEN-GOLONDRINAS                 499
dtype: int64


## **¿Cuáles son las estaciones con mayor demanda de retiros y arribos?**

Hay una coincidencia masiva entre las estaciones con más retiros y más devoluciones (los nombres que se repiten en ambos tops: Mariano Azuela, Hesíodo, Havre, Sonora, Jesús García). Esto revela zonas de alta densidad de flujos bidireccionales.

* **1. Zona Norte/Buenavista (Estaciones 548, 271-272, 555)**

Ubicaciones: Dr. Mariano Azuela (Santa María la Ribera), Jesús García - Carlos J. Meneses (atrás de la Delegación Cuauhtémoc/Buenavista), y Ribera de San Cosme e Insurgentes.

Estas estaciones son puntos clave donde la gente cambia de transporte. Están conectadas directamente con la terminal del Tren Suburbano Buenavista, el Metro Buenavista/San Cosme, y las líneas 1, 3 y 4 del Metrobús. El volumen gigantesco (especialmente la 271-272 con más de 13,500 arribos) demuestra que los usuarios utilizan ECOBICI para el último tramo de su viaje matutino desde el Estado de México o el norte de la ciudad hacia sus lugares de trabajo y/o estudio, y viceversa por la tarde.

* **2. El Corredor Corporativo (Estaciones 208, 027, 014, 031, 018):**

Ubicaciones: Hesíodo-Lamartine (Polanco), Reforma-Havre (Juárez), Reforma-Río de la Plata (frente a la Torre Mayor/Bancomer), Hamburgo-Insurgentes (Zona Rosa).

Estas estaciones representan que el destino final son las oficinas. Polanco y el Corredor Paseo de la Reforma atraen miles de viajes diarios de personas que se bajan del transporte público público e ingresan a las zonas corporativas en bicicleta.

* **3. Zona Residencial y turística (Estación 064):**

Ubicaciones: Sonora - Ámsterdam (Hipódromo/Condesa).

Funciona tanto como origen de personas que viven ahí y van a trabajar, como destino turístico, gastronómico y recreativo a cualquier hora del día.

## **¿Cuáles son las estaciones con menor demanda de retiros y arribos?**

* Coyoacán Sur e Infraestructura Nueva (Estaciones 687, 664, 666, 662, 665, 707, 694, 650):

Ubicaciones: Francisco Sosa - Panzacola (Barrio de Santa Catarina), Hortensia, Moras, Manzano, Irlanda, Centenario, Xicoténcatl. Todas estas pertenecen a la expansión de ECOBICI hacia la Alcaldía Coyoacán.

Análisis: El bajo volumen (200 a 400 viajes totales en un mes) puede deberse a tres factores:

* 1. Fecha de inauguración: De acuerdo con la página oficial de ECOBICI, (https://datos.cdmx.gob.mx/dataset/ubicacion-de-estaciones-de-ecobici-sistema-anterior/resource/c559a064-4a52-4aed-8b5a-a106b3a8f595) en febrero de 2023 existían 480 estaciones, sin embargo,  de acuerdo con la base de datos más reciente df_estaciones, actualmente existen 711. Esto quiere decir que en los últimos tres años, se han inaugurado 231 estaciones, los ID más próximos al número 711 indican una inauguración más reciente, lo que indica que las estaciones con menos movimiento aún no son tan conocidas.  

* 2. Topografía y Urbanismo: Calles como Francisco Sosa son empedradas, estrechas, con banquetas altas y menos amigables para el ciclista promedio en comparación con los carriles confinados de Reforma o Patriotismo.

* 3. Barreras viales: Estaciones como Manzano - Circuito Interior colindan con vías rápidas difíciles de cruzar en bicicleta.


In [44]:
df_clean = df_clean.rename(columns={
    "Ciclo_Estacion_Retiro": "Estacion_Retiro",
    "Ciclo_EstacionArribo": "Estacion_Arribo"
})

In [60]:
# Contamos el número de retiros y arribos en cada estación
conteo_retiros = df_clean["Estacion_Retiro"].value_counts()
conteo_arribos = df_clean["Estacion_Arribo"].value_counts()

# Creamos un DataFrame para comparar ambos conteos
df_balances = pd.DataFrame({
    "Retiros": conteo_retiros,
    "Arribos": conteo_arribos
})

# Si existe un Nan lo llenamos con 0
df_balances = df_balances.fillna(0)

# Calculamos la diferencia (Si el resultado es positivo, existen más devoluciones que retiros; si es negativo, existen más retiros que devoluciones.)
df_balances["Balance_Neto"] = df_balances["Arribos"] - df_balances["Retiros"]

# Calculamos el balance. El balance perfecto sería 1, ya que implicaría que por cada bici retirada, hay una que se devuelve a la misma estación (entre más se aleje el resultado de 1, mayor diferencia entre retiros y devoluciones existe)
df_balances["Proporcion_Arribo_Retiro"] = df_balances["Arribos"] / df_balances["Retiros"]

# Creamos una tabla de catálogo que asocie cada ID de estación con su dirección única (eliminando los duplicados)
catalogo_estaciones = df_clean[['Estacion_Retiro', 'direccion_origen']].drop_duplicates()

# Juntamos la tabla de balances con el catálogo usando el ID de la estación
df_balances = pd.merge(df_balances, catalogo_estaciones, left_on=df_balances.index, right_on='Estacion_Retiro')

# Colocamos el ID de la estación como el índice principal de nuestra tabla de balances
df_balances = df_balances.set_index('Estacion_Retiro')

# Filtramos estaciones con más de 500 retiros para prescindir de las estaciones menos utilizadas
estaciones_activas = df_balances[df_balances["Retiros"] > 500]
# Definimos las columnas que queremos mostrar en el reporte final (ahora usando 'direccion_origen')
columnas_reporte = ["direccion_origen", "Retiros", "Arribos", "Balance_Neto", "Proporcion_Arribo_Retiro"]

print("Las 5 estaciones con mayor excedente de devoluciones (constantemente llenas):")
# Mostramos las columnas clave ordenadas de mayor a menor balance
display(estaciones_activas.sort_values(by="Balance_Neto", ascending=False)[columnas_reporte].head(5))

print("\nLas 5 estaciones con mayor excedente de retiros (constantemente vacías):")
# Ordenadas de menor a mayor balance (los más negativos)
display(estaciones_activas.sort_values(by="Balance_Neto", ascending=True)[columnas_reporte].head(5))

Las 5 estaciones con mayor excedente de devoluciones (constantemente llenas):


,direccion_origen,Retiros,Arribos,Balance_Neto,Proporcion_Arribo_Retiro
Estacion_Retiro,,,,,
271-272,JESÚS GARCÍA - CARLOS J. MENESES,7711,13536,5825,1.76
266-267,JESÚS GARCÍA - CARLOS J. MENESES,2741,5977,3236,2.18
548,DR. MARIANO AZUELA - JOSE ANTONIO ALZATE,10359,12267,1908,1.18
273-274,LUIS DONALDO COLOSIO - AV. JESÚS GARCÍA,4034,5479,1445,1.36
014,REFORMA - RÍO DE PLATA,6736,8039,1303,1.19



Las 5 estaciones con mayor excedente de retiros (constantemente vacías):


,direccion_origen,Retiros,Arribos,Balance_Neto,Proporcion_Arribo_Retiro
Estacion_Retiro,,,,,
208,HESIODO-LAMARTINE,9934,8300,-1634,0.84
237-238,ANDRÉS BELLO-GEORGE ELIOT,5726,4597,-1129,0.80
211,ISAAC NEWTON-HORACIO,3848,2951,-897,0.77
210,EUGENIO SUE-HORACIO,2342,1464,-878,0.63
495,AMADO NERVO - DE LOS MAESTROS,5141,4267,-874,0.83


## **Estaciones no balanceadas: estaciones críticas con acumulación y desbasto**

1. Excedente de arrivos (abastecimiento excesivo)

En este rubro, la zona de Buenavista presenta la saturación más severa de todo el sistema:

El Nodo Jesús García - Carlos J. Meneses: Las estaciones 271-272 y 266-267 lideran el exceso de arribos. El caso de la estación 266-267 es el más alarmante, registrando una proporción de 2.18, lo que significa que se devuelven más del doble de bicicletas de las que se retiran. En conjunto, ambas estaciones acumularon un excedente neto de 9,061 bicicletas que ingresaron y posiblemente requirieron remoción física del sistema ECOBICI.

Este comportamiento se extiende a estaciones cercanas como Dr. Mariano Azuela - José Antonio Alzate (ID 548) con 1,908 arribos en exceso y Luis Donaldo Colosio - Av. Jesús García (ID 273-274) con un excedente de 1,445. Fuera de este sector, solo resalta la estación de Reforma - Río de Plata (ID 014) con un excedente de 1,303 arribos, ligada al flujo corporativo.

Impacto Operativo: Una estación desbordada impide que los usuarios concluyan su viaje, obligándolos a buscar anclajes libres en estaciones aledañas, lo que genera molestia, pérdida de tiempo y esfuerzo, y en ocasiones penalizaciones injustificadas si no encuentran espacio antes del tiempo límite de uso. Para el operador, esto exige una logística costosa de "rebalanceo vehicular" mediante camionetas dedicadas exclusivamente a vaciar estos puntos críticos durante las horas pico de llegada.

2. Estaciones con mayor excedente de retiros (desbasto constante)
La dinámica inversa se concentra de forma casi exclusiva en la zona corporativa y residencial de Polanco, donde la demanda de salidas supera por completo la velocidad de retorno.

La estación Hesíodo-Lamartine (ID 208) encabeza la lista con un déficit neto de -1,634 bicicletas (proporción de 0.84), seguida muy de cerca por Andrés Bello - George Eliot (ID 237-238) con -1,129 unidades de balance.

Las estaciones Eugenio Sue - Horacio (ID 210) e Isaac Newton - Horacio (ID 211) muestran las proporciones más bajas del análisis (0.63 y 0.77 respectivamente). En la estación 210, por cada 100 bicicletas que los usuarios retiran, solo se devuelven 63 de forma orgánica.

Impacto Operativo: Estas estaciones tienden a quedarse sin unidades disponibles con extrema rapidez, quedando inhabilitadas para los habitantes o trabajadores de la zona que buscan iniciar un trayecto, quienes tienen que buscar estaciones cercanas o bien un medio de transporte diferente.

# **Análisis del tiempo de los trayectos**

In [86]:
# 1. Convertimos primero las columnas de hora a un formato temporal reconocido por Pandas
df_clean["Hora_Retiro"] = pd.to_datetime(df_clean["Hora_Retiro"], errors='coerce')
df_clean["Hora_Arribo"] = pd.to_datetime(df_clean["Hora_Arribo"], errors='coerce')

# Combinamos la fecha con la hora extraída correctamente usando .dt
df_clean["Inicio_Viaje"] = (
    df_clean["Fecha_Retiro"].astype(str)
    + " "
    + df_clean["Hora_Retiro"].dt.strftime("%H:%M:%S")
)
df_clean["Inicio_Viaje"] = pd.to_datetime(df_clean["Inicio_Viaje"])

# ------------------------------------------------------------------------------
df_clean["Fin_Viaje"] = (
    df_clean["Fecha_Arribo"].astype(str)
    + " "
    + df_clean["Hora_Arribo"].dt.strftime("%H:%M:%S")
)
df_clean["Fin_Viaje"] = pd.to_datetime(df_clean["Fin_Viaje"])

In [63]:
# Crear columna de duración

df_clean["Duracion_Minutos"] = (
    df_clean["Fin_Viaje"] - df_clean["Inicio_Viaje"]
).dt.total_seconds() / 60

In [64]:
df_clean["Dia_Semana"] = df_clean["Inicio_Viaje"].dt.day_name() #Crear columna Dia_Semana

In [66]:
print(df_clean.columns) #revisar columnas

df_clean[["Inicio_Viaje", "Fin_Viaje", "Duracion_Minutos", "Dia_Semana"]].dtypes #revisar tipo de las nuevas columnas creadas

Index(['Genero_Usuario', 'Edad_Usuario', 'Bici', 'Estacion_Retiro',
       'Fecha_Retiro', 'Hora_Retiro', 'Estacion_Arribo', 'Fecha_Arribo',
       'Hora_Arribo', 'Edad_Atipica', 'Hora_Entera', 'direccion_origen',
       'direccion_arribo', 'Inicio_Viaje', 'Fin_Viaje', 'Duracion_Minutos',
       'Dia_Semana'],
      dtype='object')


,0
Inicio_Viaje,datetime64[ns]
Fin_Viaje,datetime64[ns]
Duracion_Minutos,float64
Dia_Semana,object


In [67]:
df_clean.head(10) #Revisar los primeros 10 registros con las nuevas columnas

,Genero_Usuario,Edad_Usuario,Bici,Estacion_Retiro,Fecha_Retiro,Hora_Retiro,Estacion_Arribo,Fecha_Arribo,Hora_Arribo,Edad_Atipica,Hora_Entera,direccion_origen,direccion_arribo,Inicio_Viaje,Fin_Viaje,Duracion_Minutos,Dia_Semana
0,M,30,4933404,059,2025-12-31,2026-07-11 23:25:20,565,2026-01-01,2026-07-11 00:00:16,False,23,POPOCATÉPETL - ÁMSTERDAM,DR. MARTINEZ DEL RIO - DR. LUCIO,2025-12-31 23:25:20,2026-01-01 00:00:16,34.93,Wednesday
1,M,27,4162376,180,2025-12-31,2026-07-11 23:43:53,672,2026-01-01,2026-07-11 00:00:31,False,23,QUINTANA ROO-VILLAHERMOSA,JOSÉ MARÍA VELASCO - DAMAS,2025-12-31 23:43:53,2026-01-01 00:00:31,16.63,Wednesday
2,M,29,3749272,014,2025-12-31,2026-07-11 23:52:15,030,2026-01-01,2026-07-11 00:00:50,False,23,REFORMA - RÍO DE PLATA,HAMBURGO - AMBERES,2025-12-31 23:52:15,2026-01-01 00:00:50,8.58,Wednesday
3,M,44,5680772,014,2025-12-31,2026-07-11 23:46:43,145,2026-01-01,2026-07-11 00:00:53,False,23,REFORMA - RÍO DE PLATA,YUCATÁN-CHIAPAS,2025-12-31 23:46:43,2026-01-01 00:00:53,14.17,Wednesday
4,M,22,5128459,031,2025-12-31,2026-07-11 23:39:58,155,2026-01-01,2026-07-11 00:01:05,False,23,HAMBURGO - INSURGENTES,JALAPA-TLAXCALA,2025-12-31 23:39:58,2026-01-01 00:01:05,21.12,Wednesday
5,M,67,2706091,015,2025-12-31,2026-07-11 23:41:23,165,2026-01-01,2026-07-11 00:01:15,False,23,REFORMA - RÍO MISSISSIPPI,BAJÍO-EJE 2 PTE. MONTERREY,2025-12-31 23:41:23,2026-01-01 00:01:15,19.87,Wednesday
6,M,33,6206304,041,2025-12-31,2026-07-11 23:46:39,027,2026-01-01,2026-07-11 00:01:18,False,23,REFORMA - AV. DE LA REPÚBLICA,REFORMA- HAVRE,2025-12-31 23:46:39,2026-01-01 00:01:18,14.65,Wednesday
7,M,34,7702525,538,2025-12-31,2026-07-11 23:48:04,027,2026-01-01,2026-07-11 00:01:37,False,23,NOGAL - ELIGIO ANCONA,REFORMA- HAVRE,2025-12-31 23:48:04,2026-01-01 00:01:37,13.55,Wednesday
8,M,28,7849797,538,2025-12-31,2026-07-11 23:47:55,027,2026-01-01,2026-07-11 00:01:37,False,23,NOGAL - ELIGIO ANCONA,REFORMA- HAVRE,2025-12-31 23:47:55,2026-01-01 00:01:37,13.70,Wednesday
9,M,32,4707477,328,2025-12-31,2026-07-11 23:46:28,022,2026-01-01,2026-07-11 00:01:46,False,23,AVENIDA 1-CALLE 7,REFORMA - MANCHESTER,2025-12-31 23:46:28,2026-01-01 00:01:46,15.30,Wednesday


**Duración promedio de los trayectos**

Se realizó un análisis descriptivo de la variable Duracion_Minutos con el objetivo de conocer el comportamiento general de los recorridos realizados por los usuarios de ECOBICI. Inicialmente, se identificó que la duración promedio de los viajes fue de 15.71 minutos, mientras que la mediana fue de 11.83 minutos, lo que indica que la mayoría de los trayectos tienen una duración relativamente corta. Sin embargo, se detectó una diferencia importante entre ambas medidas debido a la presencia de valores extremos, ya que el tiempo máximo registrado alcanzaba 257,146.30 minutos, un valor que no corresponde con el comportamiento esperado de un sistema de bicicletas compartidas y que puede afectar la interpretación del promedio.

Para obtener una representación más cercana a los patrones habituales de movilidad, se aplicó un filtro considerando únicamente viajes con una duración menor o igual a 90 minutos. Después de este ajuste, se conservaron 1,522,946 registros, manteniendo el 99.85% de los viajes originales. La nueva distribución mostró una duración promedio de 14.90 minutos y una mediana de 11.82 minutos, con un valor máximo de 89.98 minutos. Esto confirma que la mayoría de los usuarios realizan trayectos cortos, concentrados principalmente alrededor de los primeros 20 minutos.

In [68]:
print(f"Promedio de los trayectos en minutos: {df_clean['Duracion_Minutos'].mean():.2f}") # promedio de los recorridos en minutos

print(f"Mediana de los trayectos en minutos: {df_clean['Duracion_Minutos'].median():.2f}") # mediana de los recorridos en minutos

print(f"Tiempo mínimo de uso en minutos: {df_clean['Duracion_Minutos'].min():.2f}") #Tiempo mínimo de uso en minutos

print(f"Tiempo máximo de uso en minutos: {df_clean['Duracion_Minutos'].max():.2f}") #Tiempo máximo de uso en minutos


Promedio de los trayectos en minutos: 15.71
Mediana de los trayectos en minutos: 11.83
Tiempo mínimo de uso en minutos: 0.23
Tiempo máximo de uso en minutos: 257146.30


In [69]:
#Creamos una columna de viajes_normales con duración menor a una hora para disminuir los datos atípicos
viajes_normales = df_clean[
    df_clean["Duracion_Minutos"] <= 90
]

viajes_normales["Duracion_Minutos"].describe()

,Duracion_Minutos
count,1522946.00
mean,14.90
std,10.75
min,0.23
25%,7.07
50%,11.82
75%,19.78
max,89.98


**Tiempos de recorrido atípicos**

Al analizar los registros con una duración superior a 90 minutos se identificaron 2,355 viajes atípicos. La mediana de este grupo fue de 121.28 minutos, indicando que la mayoría de los casos se encuentran alrededor de dos horas de duración. Sin embargo, la diferencia entre la media (533.98 minutos) y la mediana evidencia la presencia de valores extremos que afectan la distribución. El valor máximo registrado fue de 257,146.30 minutos, equivalente a aproximadamente 179 días, lo que sugiere posibles inconsistencias en el registro temporal del viaje. Debido a esto, estos registros fueron identificados como valores atípicos para evitar que distorsionen el análisis de los patrones habituales de movilidad.

In [70]:
porcentaje = (len(viajes_normales) / len(df_clean)) * 100

print(f"Los viajes menores a 90 minutos equivalen al {porcentaje:.2f}% de los viajes originales.")

Los viajes menores a 90 minutos equivalen al 99.85% de los viajes originales.


In [71]:
viajes_atipicos = df_clean[df_clean["Duracion_Minutos"] > 90] # Creamos la variable de viajes atípicos mayores a 90 min
viajes_atipicos["Duracion_Minutos"].describe()

,Duracion_Minutos
count,2355.00
mean,533.98
std,6995.93
min,90.03
25%,101.90
50%,121.28
75%,177.98
max,257146.30


**Ordenar de manera descendente de acuerdo a la duración del viaje**

In [72]:
df_clean.sort_values(
    "Duracion_Minutos",
    ascending=False
)

,Genero_Usuario,Edad_Usuario,Bici,Estacion_Retiro,Fecha_Retiro,Hora_Retiro,Estacion_Arribo,Fecha_Arribo,Hora_Arribo,Edad_Atipica,Hora_Entera,direccion_origen,direccion_arribo,Inicio_Viaje,Fin_Viaje,Duracion_Minutos,Dia_Semana
1329970,M,30,4852091,229,2025-08-02,2026-07-11 23:24:30,1000,2026-01-28,2026-07-11 13:10:48,False,23,CARLOS DICKENS-SÉNECA,NaN,2025-08-02 23:24:30,2026-01-28 13:10:48,257146.30,Saturday
1043979,M,45,3668118,146,2025-10-03,2026-07-11 16:36:04,1000,2026-01-22,2026-07-11 21:06:01,False,16,COAHUILA-MANZANILLO,NaN,2025-10-03 16:36:04,2026-01-22 21:06:01,160109.95,Friday
728275,F,37,8460511,296,2025-11-09,2026-07-11 16:53:59,1000,2026-01-16,2026-07-11 19:45:59,False,16,DIAGONAL SAN ANTONIO-ENRIQUE RÉBSAMEN,NaN,2025-11-09 16:53:59,2026-01-16 19:45:59,98092.00,Sunday
349714,M,24,6155329,571,2025-11-03,2026-07-11 17:36:53,625,2026-01-09,2026-07-11 16:16:22,False,17,MANUEL M. FLORES - CALZADA SAN ANTONIO ABAD,ZEMPOALA - ANGEL URRAZA,2025-11-03 17:36:53,2026-01-09 16:16:22,96399.48,Monday
1002917,M,28,7738578,192-193,2025-12-20,2026-07-11 11:57:50,1000,2026-01-22,2026-07-11 10:20:35,False,11,RUBÉN DARÍO.REFORMA,NaN,2025-12-20 11:57:50,2026-01-22 10:20:35,47422.75,Saturday
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1345812,M,30,3825536,021,2026-01-28,2026-07-11 17:09:00,022,2026-01-28,2026-07-11 17:09:21,False,17,REFORMA - DUBLÍN,REFORMA - MANCHESTER,2026-01-28 17:09:00,2026-01-28 17:09:21,0.35,Wednesday
454447,M,58,3373818,385,2026-01-12,2026-07-11 10:16:22,384,2026-01-12,2026-07-11 10:16:43,False,10,POUSSIN-EMPRESA,EMPRESA-POUSSIN,2026-01-12 10:16:22,2026-01-12 10:16:43,0.35,Monday
1512729,M,30,3966611,021,2026-01-31,2026-07-11 16:14:28,022,2026-01-31,2026-07-11 16:14:48,False,16,REFORMA - DUBLÍN,REFORMA - MANCHESTER,2026-01-31 16:14:28,2026-01-31 16:14:48,0.33,Saturday
605731,M,31,3174505,034,2026-01-14,2026-07-11 18:45:45,035,2026-01-14,2026-07-11 18:46:00,False,18,LIVERPOOL - GÉNOVA,LIVERPOOL - GÉNOVA,2026-01-14 18:45:45,2026-01-14 18:46:00,0.25,Wednesday


In [77]:
#Código para responder preguntas de investigación relacionadas a los horarios y días con mayor uso

# Horarios pico y comportamiento semanal
print("Los 5 horarios con mayor demanda (a nivel general):")
print(viajes_normales["Hora_Entera"].value_counts().head(5))

print("\nDías de la semana con mayor demanda:")
orden_dias = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
print(viajes_normales["Dia_Semana"].value_counts())


# Duración promedio
print("\n" + "-"*50)
print("Duración de viajes normales")
print(f"Duración Promedio: {viajes_normales['Duracion_Minutos'].mean():.2f} minutos")
print(f"Duración Mediana:  {viajes_normales['Duracion_Minutos'].median():.2f} minutos")


# Diferencias días laborales vs fin de semana
print("\n" + "-"*50)
print("Díaslaborales VS Fines de semana")

viajes_normales = viajes_normales.copy()
viajes_normales["Tipo_Dia"] = viajes_normales["Dia_Semana"].apply(
    lambda x: "Fin de Semana" if x in ["Saturday", "Sunday"] else "Laboral"
)

# Comparación de horarios pico según el tipo de día
print("Las 4 horas pico en días laborales (Lunes a Viernes):")
print(viajes_normales[viajes_normales["Tipo_Dia"] == "Laboral"]["Hora_Entera"].value_counts().head(4))

print("\nLas 4 horas pico en fin de semana (Sábado y Domingo):")
print(viajes_normales[viajes_normales["Tipo_Dia"] == "Fin de Semana"]["Hora_Entera"].value_counts().head(4))

Los 5 horarios con mayor demanda (a nivel general):
Hora_Entera
18    133754
17    123290
15    109842
14    108698
8     107874
Name: count, dtype: int64

Días de la semana con mayor demanda:
Dia_Semana
Friday       263323
Thursday     252328
Wednesday    241894
Tuesday      236929
Monday       225227
Saturday     166600
Sunday       136645
Name: count, dtype: int64

--------------------------------------------------
Duración de viajes normales
Duración Promedio: 14.90 minutos
Duración Mediana:  11.82 minutos

--------------------------------------------------
Díaslaborales VS Fines de semana
Las 4 horas pico en días laborales (Lunes a Viernes):
Hora_Entera
18    116288
17    104080
8      95611
19     88851
Name: count, dtype: int64

Las 4 horas pico en fin de semana (Sábado y Domingo):
Hora_Entera
13    27831
12    27618
11    26018
14    25690
Name: count, dtype: int64


## **Análisis de Patrones Temporales y Ciclos de Demanda**

**Patrones de Demanda Temporal y Rutinas de Movilidad**
El análisis de los horarios muestra que el uso de ECOBICI depende completamente de las rutinas de trabajo en la Ciudad de México. Por esta razón, la actividad tiene dos momentos de máxima demanda al día (las horas pico) y disminuye notablemente durante los fines de semana.

**Distribución de la Demanda por Día de la Semana**
Al ordenar los viajes de mayor a menor volumen, se observa que la actividad del sistema se concentra de manera masiva de lunes a viernes, registrando sus niveles más altos a mitad de la semana (miércoles y jueves). Esto confirma que el servicio opera primordialmente como un componente del transporte pendular diario (viajes casa-trabajo-casa). En contraste, al llegar el fin de semana la demanda experimenta una caída notoria, tocando su punto más bajo el domingo. Esta desaceleración dominical refleja el cese de las actividades de oficina y el cambio hacia una dinámica de movilidad puramente local o de esparcimiento.

**Horarios Pico y el Impacto de la Rutina Laboral**
A nivel general, el volumen de viajes se concentra de forma crítica en dos franjas horarias que coinciden con el inicio y término de las jornadas habituales:

* El Pico Matutino (08:00 AM): Representa el primer gran flujo de usuarios, asociado a la entrada a oficinas y centros escolares, sirviendo como una herramienta clave del último tramo para conectar terminales de transporte masivo (metro, metrobús) con los destinos finales.

* El Pico Vespertino (06:00 PM / 18:00 horas): El volumen de viajes en esta franja supera consistentemente al matutino, lo que demuestra que el retorno de la población hacia los nodos de transporte o sus hogares ocurre de manera más síncrona y masiva que la dispersión de la mañana.

**Contraste entre Días Laborales y Fines de Semana**
La naturaleza del servicio cambia drásticamente al analizar el comportamiento de los usuarios según el tipo de día:

* Días Laborales (Lunes a Viernes): Se mantiene un patrón estricto de dos picos de actividad (08:00 AM y 06:00 PM). Aquí, lo más importante para los usuarios es la rapidez y la puntualidad de sus trayectos.

* Fines de Semana (Sábado y Domingo): El patrón de las horas pico desaparece. En su lugar, los viajes se distribuyen de forma suave y constante a lo largo del mediodía y la tarde (principalmente entre las 11:00 AM y las 03:00 PM). Esto demuestra que los usuarios ya no dependen de horarios fijos, transformando a ECOBICI de un transporte masivo para el trabajo en una opción recreativa, comercial y turística.

# **Perfil de las y los usuarios**

In [83]:
# Perfil de usuarios

print(df_clean[df_clean["Edad_Atipica"] == False]["Edad_Usuario"].describe())
print("-"*100)
# Filtramos para analizar a los usuarios comunes (sin las edades atípicas)
usuarios_habituales = df_clean[df_clean["Edad_Atipica"] == False]

print(f"Edad promedio de los usuarios: {usuarios_habituales['Edad_Usuario'].mean():.1f} años")
print(f"Mediana de la edad (el 50% tiene menos de): {usuarios_habituales['Edad_Usuario'].median():.1f} años")
print("-"*100)

print("\nGenero de los usuarios")

print(usuarios_habituales["Genero_Usuario"].value_counts(normalize=True) * 100)

count   1525280.00
mean         34.57
std           9.81
min          16.00
25%          28.00
50%          33.00
75%          40.00
max          87.00
Name: Edad_Usuario, dtype: float64
----------------------------------------------------------------------------------------------------
Edad promedio de los usuarios: 34.6 años
Mediana de la edad (el 50% tiene menos de): 33.0 años
----------------------------------------------------------------------------------------------------

Genero de los usuarios
Genero_Usuario
Masculino         67.53
Femenino          27.40
No especificado    3.27
Otros              1.80
Name: proportion, dtype: float64


## **Perfil General y Participación por Género**

El análisis de los datos muestra una clara diferencia entre hombres y mujeres en el uso de ECOBICI. Los hombres realizan la gran mayoría de los viajes, concentrando el 67.5% del total de trayectos (lo que equivale a más de un millón de registros). Por otro lado, las mujeres representan el 27.4% de la actividad, seguidas por un 3.3% de usuarios que prefirieron no especificar su género y un 1.8% en la categoría "Otros".

Esta diferencia ocurre en muchas ciudades del mundo. Estudios urbanos y reportes de organismos como el Institute for Transportation and Development Policy (ITDP México) y el Banco Interamericano de Desarrollo (BID) señalan que las mujeres suelen usar menos la bicicleta pública cuando perciben falta de seguridad en las calles.

* Grupos de mayor participación y edad dominante
La gran mayoría de los viajes son realizados por personas en edad de trabajar. Al analizar a los usuarios habituales —dejando de lado los pocos casos aislados con edades atípicas de 90 a 100 años—, la actividad se concentra fuertemente en jóvenes y adultos. Esto confirma que el sistema funciona principalmente como una herramienta diaria para la población activa que se desplaza hacia sus trabajos o centros de estudio.



# **Conclusiones y propuestas de acción:**

Gestión de Estaciones y Optimización de la Flota
Redistribución Dinámica Proactiva: Implementar un sistema de redistribución de unidades más intenso y frecuente justo antes de las 08:00 AM en los nodos de transporte masivo, y a las 05:30 PM en las zonas de oficinas. Esto garantizará la disponibilidad de bicicletas en el pico matutino y de anclajes libres en el vespertino.

Activación de Estaciones de Baja Movilidad: Para las estaciones donde las bicicletas permanecen fijas por mucho tiempo, se deben evaluar estrategias de reactivación, tales como su reubicación hacia puntos con mayor flujo comercial o el rediseño de la infraestructura ciclista a su alrededor para hacerlas más accesibles.

Infraestructura con Perspectiva de Género
Construcción de Calles Seguras: Para reducir la brecha actual (donde los hombres realizan el 67.5% de los viajes frente al 27.4% de las mujeres), es urgente diseñar cruces peatonales y ciclistas mejor iluminados, así como ciclovías completamente separadas del flujo vehicular en avenidas principales. Estudios de organismos como el ITDP y el BID respaldan que una infraestructura protegida disminuye la percepción de inseguridad vial, siendo el motor principal para que más mujeres adopten la bicicleta en su día a día.

Inclusión de Nuevos Sectores de la Población
Incentivos Económicos y Sociales: Como la participación de menores de 18 años y de personas de la tercera edad es muy baja, el sistema debería implementar tarifas preferenciales (como descuentos estudiantiles o vinculación directa con la tarjeta de INAPAM).

Talleres Comunitarios: Complementar los incentivos económicos con talleres gratuitos de ciclismo urbano comunitario facilitará que estos sectores pierdan el miedo a rodar en la ciudad y comiencen a utilizar ECOBICI para sus viajes cortos, escolares o recreativos durante los fines de semana, aprovechando que en esos días la curva de demanda es más suave y relajada.